File: 1930. Unique Length-3 Palindromic Subsequences.cpp
class Solution {
public:
    int countPalindromicSubsequence(string s) 
    {
        int n = s.size();
        unordered_map<char, int> minIdx, maxIdx;
        for(int ch = 'a'; ch <= 'z'; ch++)
            minIdx[ch] = INT_MAX, maxIdx[ch] = INT_MIN;

        for(int i = 0; i < n; i++)
        {
            minIdx[s[i]] = min(minIdx[s[i]], i);
            maxIdx[s[i]] = max(maxIdx[s[i]], i);
        }    

        int ans = 0;
        for(int ch = 'a'; ch <= 'z'; ch++)
        {
            if(minIdx[ch] == INT_MAX or minIdx[ch] == maxIdx[ch]) continue;
            unordered_set<char> seen;
            for(int i = minIdx[ch] + 1; i < maxIdx[ch]; i++)
            {
                if(seen.size() == 26) break;
                if(seen.count(s[i])) continue;
                seen.insert(s[i]);
                ans += 1;
            }
        }
        return ans;
    }
};

File: 1930. Unique Length-3 Palindromic Subsequences.py
class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        n =  len(s)
        pref, suf = {}, {}

        count = defaultdict(int)
        for i in range(n):
            pref[i] = dict(count)
            count[s[i]] += 1
            
        count = defaultdict(int)
        i = n - 1
        while i >= 0:
            suf[i] = dict(count)
            count[s[i]] += 1
            i -= 1
        
        store = set()
        for i in range(n):
            pf, sf = pref[i], suf[i]
            for ch in ascii_lowercase:
                if ch in pf and pf[ch] and ch in sf and sf[ch]:
                    store.add(f"{ch}{s[i]}{ch}")
        return len(store)